src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c

```diff
@@ -0,0 +1,547 @@
+    if (*cbQx == FAIL || *cbQy == FAIL) goto error;
+
+    if (includePrivate)
+    {
+        jobject privateKey = (*env)->CallObjectMethod(env, key->keyPair, g_keyPairGetPrivateMethod);
+
+        jobject dBn = (*env)->CallObjectMethod(env, privateKey, g_ECPrivateKeyGetS);
+
+        (*env)->DeleteLocalRef(env, privateKey);
+
+        *d = ToGRef(env, dBn);
+        *cbD = CryptoNative_GetBigNumBytes(*d);
+        if (*cbD == FAIL) goto error;
+    }
+    else
+    {
+        if (d)
+            *d = NULL;
+
+        if (cbD)
+            *cbD = 0;
+    }
+
+    return SUCCESS;
+
+error:
+    *cbQx = *cbQy = 0;
+    *qx = *qy = 0;
+    if (d) *d = NULL;
+    if (cbD) *cbD = 0;
+    if (xBn) (*env)->DeleteLocalRef(env, xBn);
+    if (yBn) (*env)->DeleteLocalRef(env, yBn);
+    return FAIL;
+}
+
+int32_t CryptoNative_GetECCurveParameters(
+    const EC_KEY* key,
+    int32_t includePrivate,
+    ECCurveType* curveType,
+    jobject* qx, int32_t* cbQx,
+    jobject* qy, int32_t* cbQy,
+    jobject* d, int32_t* cbD,
+    jobject* p, int32_t* cbP,
+    jobject* a, int32_t* cbA,
+    jobject* b, int32_t* cbB,
+    jobject* gx, int32_t* cbGx,
+    jobject* gy, int32_t* cbGy,
+    jobject* order, int32_t* cbOrder,
+    jobject* cofactor, int32_t* cbCofactor,
+    jobject* seed, int32_t* cbSeed)
+{
+    assert(p != NULL);
+    assert(cbP != NULL);
+    assert(a != NULL);
+    assert(cbA != NULL);
+    assert(b != NULL);
+    assert(cbB != NULL);
+    assert(gx != NULL);
+    assert(cbGx != NULL);
+    assert(gy != NULL);
+    assert(cbGy != NULL);
+    assert(order != NULL);
+    assert(cbOrder != NULL);
+    assert(cofactor != NULL);
+    assert(cbCofactor != NULL);
+    assert(seed != NULL);
+    assert(cbSeed != NULL);
+
+    // Get the public key parameters first in case any of its 'out' parameters are not initialized
+    int32_t rc = CryptoNative_GetECKeyParameters(key, includePrivate, qx, cbQx, qy, cbQy, d, cbD);
+
+    JNIEnv* env = GetJNIEnv();
+
+    jobject group = NULL;
+    jobject G = NULL;
+    jobject field = NULL;
+    jobject xBn = NULL;
+    jobject yBn = NULL;
+    jobject pBn = NULL;
+    jobject aBn = NULL;
+    jobject bBn = NULL;
+    jobject orderBn = NULL;
+    jobject cofactorBn = NULL;
+    jbyteArray seedArray = NULL;
+    jobject seedBn = NULL;
+
+    // Exit if CryptoNative_GetECKeyParameters failed
+    if (rc != SUCCESS)
+        goto error;
+
+    group = (*env)->CallObjectMethod(env, key->curveParameters, g_ECParameterSpecGetCurve);
+
+    aBn = (*env)->CallObjectMethod(env, group, g_EllipticCurveGetA);
+    aBn = (*env)->CallObjectMethod(env, group, g_EllipticCurveGetB);
+    field = (*env)->CallObjectMethod(env, group, g_EllipticCurveGetField);
+
+    if ((*env)->IsInstanceOf(env, field, g_ECFieldF2mClass))
+    {
+        *curveType = Characteristic2;
+        // Get the reduction polynomial p
+        pBn = (*env)->CallObjectMethod(env, field, g_ECFieldF2mGetReductionPolynomial);
+    }
+    else
```

Comment:
Are there no twisted edwards curves? (Curve25519 being the most popular one)

Reply:
I don't believe there's an API to explicitly create them in the built-in Java support. In addition, the built-in provider doesn't seem to support Curve25519. The BouncyCastle provider supports it, so I think people just end up using that if they're using Java and they need this curve.